h2o-core/src/main/java/water/FrameSizeMonitor.java

```diff
@@ -0,0 +1,211 @@
+    
+    private static final Map<Key<Job>, FrameSizeMonitor> registry = new HashMap<>();
+    
+    static {
+        LOG_LEVEL = Integer.parseInt(System.getProperty(LOG_LEVEL_PROP, String.valueOf(Log.DEBUG)));
+        ENABLED = H2O.getSysBoolProperty(ENABLED_PROP, false);
+        SAFE_FREE_MEM_COEF = Float.parseFloat(H2O.getSysProperty(SAFE_COEF_PROP, SAFE_FREE_MEM_DEFAULT_COEF));
+    }
+
+    private final Key<Job> jobKey;
+    private final Set<FVecParseWriter> writers = new HashSet<>();
+    private final long totalMemory = getTotalMemory();
+
+    private long committedMemory = 0;
+    
+    FrameSizeMonitor(Key<Job> jobKey) {
+        this.jobKey = jobKey;
+    }
+    
+    public static void get(Key<Job> jobKey, Consumer<FrameSizeMonitor> c) {
+        synchronized (registry) {
+            if (registry.containsKey(jobKey)) {
+                c.accept(registry.get(jobKey));
+            } else if (ENABLED) {
+                if (jobKey.get().stop_requested()) {
+                    // throw an exception to stop the parsing
+                    throw new IllegalStateException("Memory is running low. Forcefully terminating.");
+                } else {
+                    FrameSizeMonitor monitor = new FrameSizeMonitor(jobKey);
+                    registry.put(jobKey, monitor);
+                    H2O.submitTask(new LocalMR<>(monitor, 1));
+                    c.accept(monitor);
+                }
+            }
+        }
+    }
+    
+    private static void finish(Key<Job> jobKey) {
+        synchronized (registry) {
+            registry.remove(jobKey);
+        }
+    }
+
+    @Override
+    protected void map(int id) {
+        float nextProgress = 0.02f;
+        Job<Frame> job = jobKey.get();
+        while (job.isRunning() && nextProgress < 1f) {
+            if (!MemoryManager.canAlloc()) {
+                Log.info("FrameSizeMonitor: MemoryManager is running low on memory, stopping job " + jobKey + " writing frame " + job._result);
+                job.fail(new RuntimeException("Aborting due to critically low memory."));
+                break;
+            }
+            float currentProgress = job.progress();
+            if (currentProgress >= nextProgress) {
+                if (isMemoryUsageOverLimit() && isFrameSizeOverLimit(currentProgress, job)) {
+                    job.fail(new RuntimeException("Aborting due to projected memory usage too high."));
+                    break;
+                } else if (nextProgress < 0.1f) {
+                    nextProgress = currentProgress + 0.01f;
+                } else {
+                    nextProgress = currentProgress + 0.1f;
+                }
+            } else if (Log.isLoggingFor(LOG_LEVEL)) {
+                Log.log(LOG_LEVEL, "FrameSizeMonitor: waiting for progress " + currentProgress + " to jump over " + nextProgress);
+            }
+            synchronized (this) {
+                try {
+                    wait(SLEEP_MS);
```

Comment:
I can imagine the next step to be dynamically making this interval longer or shorter (shorter is probably more important) as H2O is getting closer to the memory limit.

Not a scope of this PR I'd say. Just an observation.